### PR TITLE
feat(devx): introduce storybook

### PR DESCRIPTION
**Pull request by lordvlad**
First shot at introducing storybook, with stories for the keycloak
theme pages already in the repo.

To fire up storybook run `yarn` then `yarn build-keycloak-them` then  `yarn storybook`.

### CODE REVIEW

File: .storybook/main.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module.exports = {
+  "stories": [
+    "../src/keycloak-theme/pages/Login.stories.tsx",
```
Comment (lordvlad) [inline .storybook/main.js:3]: for some reason, the usual 'glob' approach `../src/**/*stories.tsx` did not work for me, need to investigate  this

**Comment by garronej** [comment]
Hey @lordvlad,  

I'm so hype about this,  
Sorry I am completely under water, I have no time to review RN but I will soon free up time for Keycloakify.  

[Here](https://github.com/codegouvfr/keycloakify-starter/invitations) is your full access on the Keycloakify starter repo. I trust you 100%

File: .storybook/main.js (hunk 1 above)
Comment (lordvlad) [reply .storybook/main.js]: weird powershell bug. resolved

**Comment by lordvlad** [comment]
@garronej all pages are done. With your permission I'd like to update gh-pages to show the storybook instead of the demo app. But I can do that in a subsequent PR. I'd love for you to check the results first, though. I understand you're busy, no need to rush this.

**Comment by viktorlarsson** [comment]
@lordvlad Would this enable us to view emails, login etc?

**Comment by lordvlad** [comment]
@viktorlarsson all login pages. did not the work on emails yet.

Here's what the result would look like

image

**Comment by viktorlarsson** [comment]
Yea, just tried it out, works great @lordvlad !

**Comment by lordvlad** [comment]
@garronej I did not try templating emails yet, but by the looks of it, its not as full-fledged as the theme support, right? there is no react->ftl transformation in place yet, right?

**Comment by viktorlarsson** [comment]
I don't think so @lordvlad

**Comment by garronej** [comment]
First @lordvlad thank you for your work! Having a storybook like that will be such a massive improvement for the project!  

> With your permission I'd like to update gh-pages to show the storybook instead of the demo app  

I think what we can do is put the storybook under a sub path.  http://starter.keycloakify.dev/storybook  
I think it's usefull to keep starter.keycloakify.dev as the demo App that work as a minimal working authentication setup in a SPA App.  

You can have a look at the setup of GitLanding it's like that:  
- home: https://www.gitlanding.dev/
- https://www.gitlanding.dev/storybook  

And here is the CI: https://github.com/thieryw/gitlanding/blob/41ddb7d816819915f297ab7a6b89fef999f08ac8/.github/workflows/ci.yaml#L55-L68  

The important part is the `--add` that tells to add the `storybook` directory without replacing the whole content.  

> but by the looks of it, its not as full-fledged as the theme support, right? there is no react->ftl transformation in place yet, right?  

No, I wasn't very motivated to work on this since  
I think it's best to stick with a very simple template confirmation emails.  
Mails with images and HTML are more likely to end up in the spam folder or see the delivery delayed.  
I mean I'd rather put my effort in allowing the customization of the account theme (which I'm doing today)

**Comment by lordvlad** [comment]
@garronej 👍 updated ci job. offtopic, gitlanding looks like an awesome project, ill be sure to check it out more thoroughly in the future.

> Mails with images and HTML are more likely to end up in the spam folder

Not sure if HTML makes a difference, I hardly remember a mail which did _not_ have HTML. But yeah I understand your lack of motivation in that area, email layouts are tedious to work with.

> Account theme

That sounds great, we can extend this repo once you have a few pages ready.

On another note. I did not understand yet why all pages are present in the keycloakify project too: https://github.com/InseeFrLab/keycloakify/tree/main/src/lib/pages

Would it make sense to remove those pages from keycloakify and only keep them here?

**Comment by garronej** [comment]
> gitlanding looks like an awesome project  

Cool, I'm happy to hear that :)  
Be aware though that I don't stant behing this project like I stand behind some of my other stuff.  
I think the customisability is not simple enough to acheive.  

> Not sure if HTML makes a difference, I hardly remember a mail which did not have HTML.  

Yeah, you are probably right. Even the more sober emails have links. 

> On another note. I did not understand yet why all pages are present in the keycloakify project too: https://github.com/InseeFrLab/keycloakify/tree/main/src/lib/pages  

That's interesting that you see it that way.   
It's to enable fallback.  
You would copy to your repo only the pages that you want to customize in depth. The other can be customized buy only with CSS classes or via the Template.  

https://github.com/codegouvfr/keycloakify-starter/blob/3197c9f0256599bb9e35dc940b69d90dd3c1e0a0/src/keycloak-theme/pages/Login.tsx#L1-L6

In my mind the Keycloak starter was more like a tutorial that a integral part of the project.  
It's true that it would reduce code duplication to move everything to the starter but I think it will come at the cost of forcing everyone to have a ton of pages that they haven't toched in their project.  

I like that you can simply fallback to the default for the pages that you haven't customized:  
https://github.com/codegouvfr/keycloakify-starter/blob/3197c9f0256599bb9e35dc940b69d90dd3c1e0a0/src/keycloak-theme/KcApp.tsx#L72

**Comment by lordvlad** [comment]
> I think it's nice to enable the users to only customize the Template.tsx and/or the kcProps classes. This way every Keycloakify theme looks less like a fork.

Makes total sense, I was half-expecting that explanation. Maybe we can think about a tool which would get the most up-to-date file from keycloakify and adapt it, cause now I fear the duplication will result in a strong drift. Something similar to `npx create-keycloak-email-directory`? So  you  could do `npx keycloakify create idp-review-user-profile` would pull the newest idp-review-user-profile from keycloakify project.  And while we're at it, `npx keycloakify create`  could create a template just like `create-react-app`.

**Comment by garronej** [comment]
> So you could do npx keycloakify create idp-review-user-profile would pull the newest idp-review-user-profile from keycloakify project. And while we're at it, npx keycloakify create could create a template just like create-react-app.  

This would be just 😍  
If we reach that level it would be a game changer for Keycloak customizability!  
What would be also great is to be able to integrate with Nx, a significant portion of the Keycloakify's target audience uses monorepo framwork.  https://github.com/InseeFrLab/keycloakify/issues/87  

Hey, how about merging this PR? I'll maybe be dropping v7 with Account theme support tonight.  
There will be some breaking changes, the starter repo will have to be updated... well anyway it's no big deal.

**Comment by lordvlad** [comment]
> What would be also great is to be able to integrate with Nx, a significant portion of the Keycloakify's target audience uses monorepo framwork. https://github.com/InseeFrLab/keycloakify/issues/87

I have a local WIP branch which would allow the required customizability in terms of target directory etc, didnt have time to work on this, but maybe I can assign that ticket to myself, will boost motivation =]

**Comment by lordvlad** [comment]
> Hey, how about merging this PR? I'll maybe be dropping v7 with Account theme support tonight.
There will be some breaking changes, the starter repo will have to be updated... well anyway it's no big deal.

Fine by me, lemme do a rebase first

**Comment by garronej** [comment]
🎉

**Comment by lordvlad** [comment]
@garronej not sure whats failing here, maybe you can have a look? https://github.com/codegouvfr/keycloakify-starter/actions/runs/4446883177/jobs/7810023691

Is the github token only valid on tagged commits?

**Comment by garronej** [comment]
Let's goooo https://starter.keycloakify.dev/storybook 🚀

**Comment by garronej** [comment]
Hi @lordvlad,  

> So you could do npx keycloakify create idp-review-user-profile would pull the newest idp-review-user-profile from keycloakify project  

I actioned on this great idea

https://user-images.githubusercontent.com/6702424/226187260-03afe1bc-5609-4cae-8297-cb198692694c.mov

**Comment by lordvlad** [comment]
🔥

**Comment by lordvlad** [comment]
@garronej just the naming feels a bit off if i might say. Don't know if `eject` is the best word to use here 😅 
image


Might I suggest `create-keycloak-page`?

**Comment by lordvlad** [comment]
also we could also think about introducing a two layer approach already, 1 choose theme type `[email | login | account | welcome | ...?]` then choose template

**Pull request merged**